File: cobble-binding/cobble-cpp/cpp/structured/scan_plan.cpp
#include <cobble/structured/scan_plan.hpp>

#include <utility>

#include "../detail/convert.hpp"
#include "../detail/error.hpp"
#include "detail/convert.hpp"
#include "detail/impl.hpp"

namespace cobble::structured {
namespace {

rust::Vec<Byte> ToRustBytes(const std::vector<Byte> &value) {
  rust::Vec<Byte> result;
  result.reserve(value.size());
  for (const auto byte : value) result.push_back(byte);
  return result;
}

std::vector<Byte> ToBytes(const rust::Vec<Byte> &value) {
  return {value.begin(), value.end()};
}

std::vector<Byte> CopyBytes(BytesView value) {
  if (value.empty()) return {};
  return {value.begin(), value.end()};
}

structured_ffi::NativeStructuredScanSplit ToNative(const ScanSplit &value) {
  structured_ffi::NativeStructuredScanSplit result;
  result.shard = detail::ToNativeShardSnapshot(value.shard);
  result.has_start = value.start_inclusive.has_value();
  result.start = value.start_inclusive ? ToRustBytes(*value.start_inclusive)
                                       : rust::Vec<Byte>{};
  result.has_end = value.end_exclusive.has_value();
  result.end = value.end_exclusive ? ToRustBytes(*value.end_exclusive)
                                   : rust::Vec<Byte>{};
  result.has_start_after = value.start_after_exclusive.has_value();
  result.start_after_bucket =
      value.start_after_exclusive ? value.start_after_exclusive->bucket : 0;
  result.start_after_key = value.start_after_exclusive
                               ? ToRustBytes(value.start_after_exclusive->key)
                               : rust::Vec<Byte>{};
  result.has_end_at = value.end_at_inclusive.has_value();
  result.end_at_bucket =
      value.end_at_inclusive ? value.end_at_inclusive->bucket : 0;
  result.end_at_key = value.end_at_inclusive
                          ? ToRustBytes(value.end_at_inclusive->key)
                          : rust::Vec<Byte>{};
  return result;
}

ScanSplit ToSplit(const structured_ffi::NativeStructuredScanSplit &value) {
  ScanSplit result;
  result.shard = detail::ToShardSnapshot(value.shard);
  if (value.has_start) result.start_inclusive = ToBytes(value.start);
  if (value.has_end) result.end_exclusive = ToBytes(value.end);
  if (value.has_start_after)
    result.start_after_exclusive = ScanSplitBoundary{
        value.start_after_bucket, ToBytes(value.start_after_key)};
  if (value.has_end_at)
    result.end_at_inclusive =
        ScanSplitBoundary{value.end_at_bucket, ToBytes(value.end_at_key)};
  return result;
}

}  // namespace

ScanPlan::ScanPlan(GlobalSnapshot snapshot) : snapshot_(std::move(snapshot)) {}
ScanPlan ScanPlan::FromGlobalSnapshot(GlobalSnapshot snapshot) {
  return ScanPlan(std::move(snapshot));
}
ScanPlan &ScanPlan::WithStart(BytesView start_inclusive) {
  start_inclusive_ = CopyBytes(start_inclusive);
  return *this;
}
ScanPlan &ScanPlan::WithEnd(BytesView end_exclusive) {
  end_exclusive_ = CopyBytes(end_exclusive);
  return *this;
}
ScanPlan &ScanPlan::WithoutStart() noexcept {
  start_inclusive_.reset();
  return *this;
}
ScanPlan &ScanPlan::WithoutEnd() noexcept {
  end_exclusive_.reset();
  return *this;
}
std::vector<ScanSplit> ScanPlan::Splits() const {
  std::vector<ScanSplit> result;
  result.reserve(snapshot_.shards.size());
  for (const auto &shard : snapshot_.shards) {
    result.push_back(
        {shard, start_inclusive_, end_exclusive_, std::nullopt, std::nullopt});
  }
  return result;
}

std::string ScanPlan::ToJson() const {
  const auto json = detail::Translate([&] {
    return cobble::ffi::native_scan_plan_to_json(
        cobble::detail::ToNativeScanPlan(snapshot_, start_inclusive_,
                                         end_exclusive_));
  });
  return {json.data(), json.size()};
}

ScanPlan ScanPlan::FromJson(std::string_view json) {
  auto native = detail::Translate([&] {
    return cobble::ffi::native_scan_plan_from_json(detail::RustStr(json));
  });
  auto plan =
      FromGlobalSnapshot(cobble::detail::ToGlobalSnapshot(native.snapshot));
  if (native.has_start) {
    plan.start_inclusive_ =
        std::vector<Byte>(native.start.begin(), native.start.end());
  }
  if (native.has_end) {
    plan.end_exclusive_ =
        std::vector<Byte>(native.end.begin(), native.end.end());
  }
  return plan;
}

ScanSplitPartition ScanSplit::SplitAfter(BucketId bucket,
                                         BytesView key_inclusive) const {
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_scan_split_split_after(
        ToNative(*this), bucket, detail::RustBytes(key_inclusive));
  });
  if (native.size() != 2)
    throw Error(ErrorCode::kInvalidState,
                "Rust bridge returned an invalid structured split partition");
  return {ToSplit(native[0]), ToSplit(native[1])};
}

std::string ScanSplit::ToJson() const {
  const auto json = detail::Translate([&] {
    return structured_ffi::native_structured_scan_split_to_json(
        ToNative(*this));
  });
  return {json.data(), json.size()};
}

ScanSplit ScanSplit::FromJson(std::string_view json) {
  return ToSplit(detail::Translate([&] {
    return structured_ffi::native_structured_scan_split_from_json(
        detail::RustStr(json));
  }));
}

ScanCursor ScanSplit::OpenScanner(std::string_view config_json,
                                  const ScanOptions &options) const {
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ScanOptions has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_scan_split_open_scanner(
        detail::RustStr(config_json), ToNative(*this), *options.impl_->native);
  });
  return ScanCursor(std::make_unique<ScanCursor::Impl>(std::move(native)));
}

ScanCursor ScanSplit::OpenScannerFile(std::string_view config_path,
                                      const ScanOptions &options) const {
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ScanOptions has been moved from");
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_scan_split_open_scanner_file(
        detail::RustStr(config_path), ToNative(*this), *options.impl_->native);
  });
  return ScanCursor(std::make_unique<ScanCursor::Impl>(std::move(native)));
}

}  // namespace cobble::structured

File: cobble-binding/cobble-cpp/src/structured/scan_plan.rs
use std::collections::BTreeMap;
use std::ops::RangeInclusive;

use cobble_binding::Config;
use cobble_binding::structured::StructuredScanSplit;

use crate::structured_bridge::ffi;

use super::conversion::{format_error, input_error};
use super::scan::{NativeStructuredScanCursor, native_structured_scan_cursor_from_split};
use super::{BridgeResult, NativeStructuredScanOptions};

fn native_split(value: StructuredScanSplit) -> BridgeResult<ffi::NativeStructuredScanSplit> {
    let (has_start_after, start_after_bucket, start_after_key) =
        match (value.start_bucket, value.start_key_exclusive) {
            (Some(bucket), Some(key)) => (true, bucket, key),
            (None, None) => (false, 0, Vec::new()),
            _ => {
                return Err(input_error(
                    "structured split start bucket and key must be set together",
                ));
            }
        };
    let (has_end_at, end_at_bucket, end_at_key) = match (value.end_bucket, value.end_key_inclusive)
    {
        (Some(bucket), Some(key)) => (true, bucket, key),
        (None, None) => (false, 0, Vec::new()),
        _ => {
            return Err(input_error(
                "structured split end bucket and key must be set together",
            ));
        }
    };
    Ok(ffi::NativeStructuredScanSplit {
        shard: native_shard(value.shard),
        has_start: value.start.is_some(),
        start: value.start.unwrap_or_default(),
        has_end: value.end.is_some(),
        end: value.end.unwrap_or_default(),
        has_start_after,
        start_after_bucket,
        start_after_key,
        has_end_at,
        end_at_bucket,
        end_at_key,
    })
}

fn split(value: ffi::NativeStructuredScanSplit) -> BridgeResult<StructuredScanSplit> {
    Ok(StructuredScanSplit {
        shard: shard(value.shard)?,
        start: value.has_start.then_some(value.start),
        end: value.has_end.then_some(value.end),
        start_bucket: value.has_start_after.then_some(value.start_after_bucket),
        start_key_exclusive: value.has_start_after.then_some(value.start_after_key),
        end_bucket: value.has_end_at.then_some(value.end_at_bucket),
        end_key_inclusive: value.has_end_at.then_some(value.end_at_key),
    })
}

pub(crate) fn native_structured_scan_split_split_after(
    value: ffi::NativeStructuredScanSplit,
    bucket: u16,
    key: &[u8],
) -> BridgeResult<Vec<ffi::NativeStructuredScanSplit>> {
    let partition = split(value)?
        .split_after(bucket, key.to_vec())
        .map_err(format_error)?;
    Ok(vec![
        native_split(partition.before)?,
        native_split(partition.after)?,
    ])
}

pub(crate) fn native_structured_scan_split_to_json(
    value: ffi::NativeStructuredScanSplit,
) -> BridgeResult<String> {
    serde_json::to_string(&split(value)?)
        .map_err(|error| input_error(format!("cannot encode structured split JSON: {error}")))
}

pub(crate) fn native_structured_scan_split_from_json(
    json: &str,
) -> BridgeResult<ffi::NativeStructuredScanSplit> {
    let value: StructuredScanSplit = serde_json::from_str(json)
        .map_err(|error| input_error(format!("invalid structured split JSON: {error}")))?;
    native_split(value)
}

fn open_scanner(
    config: Config,
    value: ffi::NativeStructuredScanSplit,
    options: &NativeStructuredScanOptions,
) -> BridgeResult<Box<NativeStructuredScanCursor>> {
    if options.options.as_cobble().should_stop_at_block_boundary() {
        return Err(input_error(
            "stop_at_block_boundary is not supported for structured split scanners",
        ));
    }
    opendal::install_default();
    let scanner = split(value)?
        .create_scanner(config, &options.options)
        .map_err(format_error)?;
    Ok(native_structured_scan_cursor_from_split(scanner))
}

pub(crate) fn native_structured_scan_split_open_scanner(
    config_json: &str,
    value: ffi::NativeStructuredScanSplit,
    options: &NativeStructuredScanOptions,
) -> BridgeResult<Box<NativeStructuredScanCursor>> {
    open_scanner(
        Config::from_json_str(config_json).map_err(format_error)?,
        value,
        options,
    )
}

pub(crate) fn native_structured_scan_split_open_scanner_file(
    config_path: &str,
    value: ffi::NativeStructuredScanSplit,
    options: &NativeStructuredScanOptions,
) -> BridgeResult<Box<NativeStructuredScanCursor>> {
    open_scanner(
        Config::from_path(config_path).map_err(format_error)?,
        value,
        options,
    )
}

fn native_shard(value: cobble_binding::ShardSnapshotRef) -> ffi::NativeShardSnapshot {
    ffi::NativeShardSnapshot {
        ranges: value
            .ranges
            .into_iter()
            .map(|range| ffi::NativeBucketRange {
                start_inclusive: *range.start(),
                end_inclusive: *range.end(),
            })
            .collect(),
        families: value
            .column_family_ids
            .into_iter()
            .map(|(name, id)| ffi::NativeFamily { name, id })
            .collect(),
        db_id: value.db_id,
        snapshot_id: value.snapshot_id,
        manifest_path: value.manifest_path,
        timestamp_seconds: value.timestamp_seconds,
        data_size_bytes: value.data_size_bytes,
        incremental_data_size_bytes: value.incremental_data_size_bytes,
    }
}

fn shard(value: ffi::NativeShardSnapshot) -> BridgeResult<cobble_binding::ShardSnapshotRef> {
    if value.db_id.is_empty() || value.manifest_path.is_empty() {
        return Err(input_error(
            "structured split shard db_id and manifest_path must not be empty",
        ));
    }
    Ok(cobble_binding::ShardSnapshotRef {
        ranges: ranges(value.ranges)?,
        column_family_ids: families(value.families)?,
        db_id: value.db_id,
        snapshot_id: value.snapshot_id,
        manifest_path: value.manifest_path,
        timestamp_seconds: value.timestamp_seconds,
        data_size_bytes: value.data_size_bytes,
        incremental_data_size_bytes: value.incremental_data_size_bytes,
    })
}

fn ranges(values: Vec<ffi::NativeBucketRange>) -> BridgeResult<Vec<RangeInclusive<u16>>> {
    if values.is_empty() {
        return Err(input_error(
            "structured split shard ranges must not be empty",
        ));
    }
    values
        .into_iter()
        .map(|range| {
            if range.start_inclusive > range.end_inclusive {
                return Err(input_error("structured split shard range is reversed"));
            }
            Ok(range.start_inclusive..=range.end_inclusive)
        })
        .collect()
}

fn families(values: Vec<ffi::NativeFamily>) -> BridgeResult<BTreeMap<String, u8>> {
    let mut by_name = BTreeMap::new();
    let mut by_id = BTreeMap::new();
    for family in values {
        if family.name.is_empty() {
            return Err(input_error("column family name must not be empty"));
        }
        if by_name.insert(family.name.clone(), family.id).is_some() {
            return Err(input_error("duplicate column family name"));
        }
        if by_id.insert(family.id, family.name).is_some() {
            return Err(input_error("duplicate column family id"));
        }
    }
    Ok(by_name)
}
